Add Chopsticks docs (#2)
papermoonio/polkadot-ecosystem-docs-draft: docs/dev-tools/chopsticks/overview.md

!!!note
    Chopsticks currently does not support Ethereum JSON-RPC calls, so you cannot use it to fork your chain and connect Metamask.

## Prerequisites

Before you begin, ensure you have the following installed:

- [Node.js](https://nodejs.org/en/){target=_blank}
- Package manager - [npm](https://www.npmjs.com/){target=_blank} should be installed with Node.js by default. Alternatively, you can use other package managers like [Yarn](https://yarnpkg.com/){target=_blank}

## Getting Started

You can install Chopsticks either globally or locally in your project. Choose the option that best fits your development workflow.

### Global Installation

To install Chopsticks globally, allowing you to use it across multiple projects, run:

```bash
npm i -g @acala-network/chopsticks
```

> Now, you should be able to run the `chopsticks` command from your terminal.

Author: I'm not sure about this one. If you run Chopsticks without any parameters, it just hangs there and doesn't do anything. The process starts but doesn't complete or produce output. It's unclear if this is intended behavior or a bug.

> Now, you should be able to run the `chopsticks` command from your terminal.

Reviewer: I would run it with a minimal working example with Polkadot (and explain later the flag). Note that for the local installation you do state the command of running Chopsticks (with NPX). So for those things you should always add a terminal log, but also to standardize it between Global and Local, add an example here

### Local Installation

To use Chopsticks in a specific project, first create a new directory and initialize a Node.js project:

```bash
mkdir my-chopsticks-project
cd my-chopsticks-project
npm init -y
```

Then, install Chopsticks as a local dependency:

```bash

> npm i @acala-network/chopsticks

Author: I don't think we covered all the options in our previous tutorials. We could definitely do that, but which package manager should we choose? There are so many out there, like npm, yarn, bun, pnpm, etc. The process is pretty similar with any of them.

> npm i @acala-network/chopsticks

Reviewer: Yeah agreed - Unsure what is the right approach here. The only thing is that defaulting to NPM is not great IMO, but anyways.

We could define a standard and state NPM and YARN as default. I'm now using pnpm but anyways

> npm i @acala-network/chopsticks

Reviewer 2: I vote `npm`, `yarn`, and `pnpm` - I don't think we'll need more than that

```

Finally, you can run Chopsticks using the `npx` command:

```bash

> npx @acala-network/chopsticks

Reviewer: Is NPX a standard thing? Or it is something you need to install globally? If so, we need to add that here or just run the binary the non NPX way

Author: NPX comes automatically installed with NPM. In this example, they create a project from scratch and do not need to install anything else.

```

> ## Configuration

Reviewer: I would love to see this section expand a bit more. I mean you can literally overwrite a lot of things with Chopsticks configuration file, like giving a bunch of tokens to predefined accounts, or assets in AssetHub. So maybe do a bit more of a thorough explanation of the structure of a configuration file?

Author: I've actually been thinking about making a cheatsheet with tons of examples covering all that stuff.
The thing is, the config file does a bunch of different things, and what you're mentioning is mainly part of the set_storage side. I think a separate cheatsheet might be more useful thanputting it all in here.
What do you think?

Reviewer: Ok agree - if you create a separate page then I would add a link here. But I think one of the powerful things is to create your own config files in which you can do crazy things like override sudo, councils, set asset balances, etc.

> To run Chopsticks, you need to configure some parameters. This can be set either through using a configuration file or the command line interface (CLI).

Reviewer: ```suggestion
To run Chopsticks, you need to configure some parameters. You can do this using a configuration file or the command-line interface (CLI).
```

An example of a configuration file for Polkadot is as follows:

```yaml
endpoint:
  - wss://rpc.ibp.network/polkadot
  - wss://polkadot-rpc.dwellir.com
mock-signature-host: true
block: ${env.POLKADOT_BLOCK_NUMBER}
db: ./db.sqlite
runtime-log-level: 5

import-storage:
  System:
    Account:
      -
        -
          - 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY
        - providers: 1
          data:
            free: '10000000000000000000'
  ParasDisputes:
    $removePrefix: ['disputes'] # those can makes block building super slow
```

To run Chopsticks using a configuration file, utilize the `--config` flag. You can use a raw GitHub URL, a path to a local file, or simply the chain's name. For example, the following commands all use Polkadot's configuration in the same way:

=== "GitHub URL"

    ```bash
    npx @acala-network/chopsticks \
    --config=https://raw.githubusercontent.com/AcalaNetwork/chopsticks/master/configs/polkadot.yml
    ```

=== "Local File Path"

    ```bash
    npx @acala-network/chopsticks --config=configs/polkadot.yml
    ```

=== "Chain Name"

    ```bash
    npx @acala-network/chopsticks --config=polkadot
    ```

> 

Reviewer: Add an example terminal output of what happens when you run chopsticks?

!!! note
    If using a file path, make sure you've downloaded the [Polkadot configuration file](https://github.com/AcalaNetwork/chopsticks/blob/master/configs/polkadot.yml){target=_blank}, or have created your own.

### Using Command Line Interface (CLI)

Alternatively, all settings (except for genesis and timestamp) can be configured via command-line flags, providing a comprehensive method to set up the environment. For example, the following command forks Polkadot at block 100.

```bash
npx @acala-network/chopsticks \
--endpoint wss://polkadot-rpc.dwellir.com \
--block 100
```

> 

Reviewer: Add example terminal output

> ## Interacting with a Fork

Reviewer: Should we add that if you run multiple forks they are automatically running in 8000, 8001, etc? And that you can set the port you want to run the fork on with `--port`?

Author: It is already mentioned in the Configuration section. Maybe we could add the multiple forks scenario

When running a fork, it's accessible by default at:

```bash
ws://localhost:8000
```

You can interact with the forked chain using various [libraries](https://wiki.polkadot.network/docs/build-tools-index#libraries){target=_blank} such as [Polkadot.js](https://polkadot.js.org/docs/){target=_blank} and its user interface, [Polkadot.js Apps](https://polkadot.js.org/apps/#/explorer){target=_blank}.

### Using Polkadot.js Apps

To interact with Chopsticks via the hosted user interface, visit [Polkadot.js Apps](https://polkadot.js.org/apps/#/explorer){target=_blank} and follow these steps:

1. Click the network icon in the top left corner
2. Scroll to the bottom and select **Development**
3. Choose **Custom**
4. Enter `ws://localhost:8000` in the input field
4. Click the **Switch** button

> ![](/polkadot-ecosystem-docs-draft/images/dev-tools/chopsticks/chopsticks-1.webp)

Reviewer: Can we add this image just below the `1.` and then the other one where it is?

![](/polkadot-ecosystem-docs-draft/images/dev-tools/chopsticks/chopsticks-2.webp)

You should now be connected to your local fork and can interact with it as you would with a real chain.

### Using Polkadot.js Library

For programmatic interaction, you can use the Polkadot.js library. Here's a basic example:

```js
import { ApiPromise, WsProvider } from '@polkadot/api';

async function connectToFork() {
  const wsProvider = new WsProvider('ws://localhost:8000');
  const api = await ApiPromise.create({ provider: wsProvider });

>   

Reviewer: I believe it is a good practice to do `await api.isReady` -> https://polkadot.js.org/docs/api/start/create#api-instance

Author: Nice one! I didn't know about this

  // Now you can use 'api' to interact with your fork
  console.log(`Connected to chain: ${await api.rpc.system.chain()}`);
}

connectToFork();
```

## Replaying Blocks

Chopsticks allows you to replay specific blocks from a chain, which is useful for debugging and analyzing state changes.  You can use the parameters in the [Configuration](#configuration) section to set up the chain configuration, and then use the run-block subcommand with additional options:

|    Option     |           Description           |
| :-----------: | :-----------------------------: |
| `output-path` |    File path to print output    |
|    `html`     | Generate html with storage diff |
|    `open`     |       Open generated html       |

For example, to replay block 1000 from Polkadot and save the output to a JSON file:

```bash
npx @acala-network/chopsticks run-block  \
--endpoint wss://polkadot-rpc.dwellir.com  \
--output-path ./polkadot-output.json  \
--block 1000
```

> 

Reviewer: Provide Terminal output example?

## XCM Testing
To test XCM (Cross-Consensus Messaging) messages between networks, you can fork multiple parachains and a relay chain locally using Chopsticks. 

|    Option    |      Description       |
| :----------: | :--------------------: |
| `relaychain` | Relaychain config file |
| `parachain`  | Parachain config file  |

For example, to fork Moonbeam, Astar, and Polkadot enabling XCM between them, you can use the following command:

```bash
npx @acala-network/chopsticks xcm \
--r polkadot \
--p moonbeam \
--p astar
```

After running it, you should see output similar to the following:

> <div id="termynal" data-termynal>

Reviewer: Why here only and not everywhere else? haha

    <span data-ty>    app: "chopsticks"</span>
    <span data-ty>[13:46:25.144] INFO: Loading config file https://raw.githubusercontent.com/AcalaNetwork/chopsticks/master/configs/polkadot.yml</span>
        <span data-ty>    app: "chopsticks"</span>
    <span data-ty>        chopsticks::executor  TRACE: Calling Metadata_metadata</span>
    <span data-ty>        chopsticks::executor  TRACE: Completed Metadata_metadata</span>
    <span data-ty>[13:46:53.320] INFO: Polkadot RPC listening on port 8002</span>
    <span data-ty>    app: "chopsticks"</span>
    <span data-ty>[13:46:54.038] INFO (xcm): Connected relaychain 'Polkadot' with parachain 'Moonbeam'</span>
    <span data-ty>    app: "chopsticks"</span>
    <span data-ty>[13:46:55.028] INFO (xcm): Connected relaychain 'Polkadot' with parachain 'Astar'</span>
    <span data-ty>    app: "chopsticks"</span>
</div>

Now you can interact with the forked chains using the ports specified in the output.

## WebSocket Commands

Chopstick's internal WebSocket server has special endpoints that allow the manipulation of the local Substrate chain.

These are the methods that can be invoked and their parameters:

> ???+ function "**dev_newBlock** (newBlockParams) — Generates one or more new blocks"

Reviewer: Unsure about this way to add these commands. I belive @eshaben found  better way to document such things

Reviewer 2: Inside of an expandable element is correct, but it is missing the styling that we typically use, so it looks a bit different, so I'll create a separate PR for that. For a page like this, I would stick with the tabs for Parameters and Example, but I think I would drop the Parameters table for a list in this case (with styled types). 

I'll send you an example
